### PR TITLE
Restrict active submissions to one per assessment

### PR DESCRIPTION
**Pull request by lowjoel**
1. Implement manually preloading associations
 2. Implement scopes for querying assessments
 3. Changed the controller to use those scopes

This is ugly, but I can't find a cleaner way. Rails does not allow us to specify conditions on the JOIN which are parameterised and supports eager loading.

### CODE REVIEW

**Comment by lowjoel** [comment]
I think I'll refactor the manual eager loading to use the ActiveRecord API instead.

**Comment by allenwq** [comment]
Sorry, I don't understand to problem you are facing. Is that you only want to eager load the active submission and rails cannot do that by default?

**Comment by lowjoel** [comment]
If you use eager_load on an assessment, you eager load the assessment and ALL submissions, including those not belonging to the current student. Which means we end up preloading a few hundred submissions which are useless.

**Comment by allenwq** [comment]
I see.

**Comment by lowjoel** [comment]
You can google for "activerecord dynamic condition association eager load" and people have been trying to get it into Rails for a long time, and the Rails dev team thinks it's not necessary.

**Comment by lowjoel** [comment]
I've not squashed the commits yet. The new API follows ActiveRecord's, any comments?

File: app/models/course/assessment.rb [hunk 1]
```diff
@@ -12,7 +12,21 @@ class Course::Assessment < ActiveRecord::Base
...
            through: :questions, source: :actable,
            source_type: Course::Assessment::Question::MultipleResponse.name
-  has_many :submissions, dependent: :destroy
+  has_many :submissions, inverse_of: :assessment, dependent: :destroy
+
+  # @!method with_submissions_by(creator)
+  #   Includes the submissions by the provided user.
+  #   @param [User|nil] creator The creator to preload submissions by, or +nil+ to preload all
+  #     submissions.
+  scope :with_submissions_by, (lambda do |creator = nil|
```
Comment (allenwq) [inline app/models/course/assessment.rb]: I think `nil` to preload all submissions does not make sense

Comment (lowjoel) [inline app/models/course/assessment.rb]: good point, thanks!

**Comment by lowjoel** [comment]
@allenwq if this design is OK, I'll squash and this can be merged.

**Comment by allenwq** [comment]
@lowjoel Sure, I don't have any comments on the preloaded part, not familiar with it :(

**Comment by lowjoel** [comment]
@allenwq Thanks, I've squashed it.

**Pull request merged**